### ISSUE TITLE
[PATCH]t.p.formmethod.Choice.coerce should handle allowNone

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4835]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Tv reported|
|-|-|
|Trac ID|trac#353|
|Type|defect|
|Created|2003-10-19 18:47:14Z|

<details><summary>Searchable metadata</summary>

```
trac-id__353 353
type__defect defect
reporter__Tv Tv
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__web web
time__1066589234000000 1066589234000000
changetime__1468923592348201 1468923592348201
version__ 
owner__ 
cc__itamarst cc__Tv cc__dp
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Tv commented|
|-|-|

```
#!html
<pre>
Currently, Choices do not support allowNone. This should help:

Index: twisted/python/formmethod.py
===================================================================
RCS file: /cvs/Twisted/twisted/python/formmethod.py,v
retrieving revision 1.41
diff -u -u -r1.41 formmethod.py
--- twisted/python/formmethod.py	10 Oct 2003 23:08:03 -0000	1.41
+++ twisted/python/formmethod.py	19 Oct 2003 12:39:06 -0000
@@ -214,6 +214,8 @@
         Argument.__init__(self, name, default, shortDesc, longDesc, hints,
allowNone=allowNone)
 
     def coerce(self, inIdent):
+        if not inIdent.strip() and self.allowNone:
+            return None
         for ident, val, desc in self.choices:
             if ident == inIdent:
                 return val
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Tv commented|
|-|-|

```
#!html
<pre>
Note just inIdent.strip() makes unit tests fail as they pass in integer 1, using
str(inIdent).strip() does work, though.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Why is this a bug? This looks like a feature request to me.
Please downgrade.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Tagging as patch
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Problem is, allowNone is true by default, so this would probably change
behaviour.  Also as is it'll return None for any whitespace key, which are valid
choices in the dict... There's probably some solution though.

Real solution is to kill formmethod and form. I've been adding all these
features for my own use cases, but it's ending up being hacky and inconsistent
if it is going to be backwards comptable, and it's rather icky in general :(
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Tv commented|
|-|-|

```
#!html
<pre>
The reason why I see this as a bug is exactly because Choice.__init__ takes
allowNone parameter, and even has it defaulting to true. I was very surprised
that it didn't work.

How about making the default allowNone=0 and enabling the support? That wouldn't
change behaviour for any existing callers (assuming no one would pass
allowNone=1 and manage to ignore it being broken, and even then we'd just change
semantics to what they requested).

I'm not sure what should be done about the whitespace-only keys issue itamar
raised. I just copy-pasted the code from other coerces ;) I don't see it could
end up having whitespace-only inIdent except if some idiot does &lt;select name="
">, &lt;select name="  ">, etc, and that can be handled by just dropping the .strip().

Itamar, you can't just say formmethod and form need to be killed. I'm going to
have to abandon Twisted as a realistic platform if you guys keep doing this
"show the donkey the carrot, then take it away" too much. There has got to be
something _else_ in place, _then_ let's talk about obsoleting old things.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Changing allowNone to default to false for Choice would probably work. 

As for whitespace - you should only get None if *no* item was selected at all
(e.g &lt;input type="radio"> bunch where none of the radio buttons were selected),
and thus browser didn't submit any value. I'm not sure what would get passed to
coerce in that case (None? empty string?). Try to see. But just killing the
strip() would do, probably.

Having allowNone default to false here is a bit inconsistent, but it makes
sense, I guess.

As for "realistic platform" - go check out the stuff the Zope people have been
doing. They keep writing new APIs (including zope3, which is totally
incompatible) for same reason we do, they come up with better designs. We will
not be *removing* form. I depend on it too for a paying project I've been
working on for the past year. Nor will twisted.web.widgets be removed, I depend
on it too in one of my other apps. But sooner or later we will have better way
of doings things.  My point was more that for backwards compatibility reasons
there's a limit to how much form can be cleaned up.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
You should probably talk about how and when to migrate this to nevow.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Closing as wont fix, because it relates to woven which is deprecated and unmaintained.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|GitHub commented|
|-|-|

In changeset 884af50645efd22eca9012b890c7e0b90f6e2689
```
#!CommitTicketReference repository="" revision="884af50645efd22eca9012b890c7e0b90f6e2689"
Merge pull request #353 from twisted/8626-rodrigc-win32reactor-py3

Author: rodrigc
Reviewer: acabhishek942
Fixes: #8626

Fix win32reactor on Python 3
```
